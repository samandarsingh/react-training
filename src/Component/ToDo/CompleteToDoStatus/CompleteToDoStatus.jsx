import React from "react";
import PropTypes from "prop-types";
import Button from "../../Shared/Button";
/**
 * this component renders status button
 * @param {bool} completeToDoStatus 
 * @param {func} handleOnClick
 * @param {number} index
 * @returns Element
 */
const CompleteToDoStatus = props => {
    const {
        completeToDoStatus,
        handleOnClick,
        index
    } = props;
    return (
        <Button 
            buttonType="button"
            buttonName={completeToDoStatus ? "Completed":"In Progress"}
            buttonClass={completeToDoStatus ? "isCompletedButton":"statusButton"}
            handleOnClick={handleOnClick}
            index={index}
            status={completeToDoStatus ? true:false}            
        />
    );
}
CompleteToDoStatus.defaultProps = {
    completeToDoStatus: false,
    handleOnClick: () => {},
    index: 0
}
CompleteToDoStatus.propTypes = {
    completeToDoStatus: PropTypes.bool.isRequired,
    handleOnClick: PropTypes.func.isRequired,
    index: PropTypes.number.isRequired
}
export default CompleteToDoStatus;