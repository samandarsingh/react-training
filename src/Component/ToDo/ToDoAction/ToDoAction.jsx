import React from "react";
import PropTypes from "prop-types";
import Button from "../../Shared/Button";
/**
 * this component renders status button
 * @param {func} handleOnClickUpdateToDoStatus 
 * @param {func} handleOnClickRemoveToDoStatus
 * @param {number} index
 * @returns Element
 */
const ToDoAction = props => {
    const {
        handleOnClickUpdateToDoStatus,
        handleOnClickRemoveToDoStatus,
        index        
    } = props;
    return (
        <>
            <Button 
                buttonType="button"
                buttonName="Edit"
                buttonClass="secondary"
                handleOnClick={handleOnClickUpdateToDoStatus}
                index={index}            
            />             
            <Button 
                buttonType="button"
                buttonName="Delete"
                buttonClass="error" 
                handleOnClick={handleOnClickRemoveToDoStatus}    
                index={index}                            
            /> 
        </>
    );
}
ToDoAction.defaultProps = {
    handleOnClickUpdateToDoStatus: () => {},
    handleOnClickRemoveToDoStatus: () => {},
    index: 0
}
ToDoAction.propTypes = {
    handleOnClickUpdateToDoStatus: PropTypes.func.isRequired,
    handleOnClickRemoveToDoStatus: PropTypes.func.isRequired,
    index: PropTypes.number.isRequired
}
export default ToDoAction;