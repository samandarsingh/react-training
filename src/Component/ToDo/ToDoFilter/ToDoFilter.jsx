import React from "react";
import PropTypes from "prop-types";
import styles from "./ToDoFilter.module.scss";
import Button from "../../Shared/Button";
/**
 * 
 * @param {fun} handleOnClick
 * @param {boolean} activeFilter
 * @returns Element
 */
const ToDoFilter = props => {
    const {handleOnClick, activeFilter} =props;
    return (
        <div className={styles.todoFilterContainer}>
            <Button 
                buttonType="button"
                buttonName="All"
                buttonClass={ activeFilter===0 ? "isCompletedButton":"statusButton"}
                handleOnClick={handleOnClick}    
                index={0} 
                status={ activeFilter===0 ? true : false} 
                disabled={ activeFilter===0 ? true : false}                        
            />
            &nbsp;
            <Button 
                buttonType="button"
                buttonName="Active"
                buttonClass={ activeFilter===1 ? "isCompletedButton":"statusButton"}
                handleOnClick={handleOnClick}    
                index={1} 
                status={ activeFilter===1 ? true : false} 
                disabled={ activeFilter===1 ? true : false}                       
            />
            &nbsp;
            <Button 
                buttonType="button"
                buttonName="Completed"
                buttonClass={ activeFilter===2 ? "isCompletedButton":"statusButton"}
                handleOnClick={handleOnClick}    
                index={2} 
                status={ activeFilter===2 ? true : false} 
                disabled={ activeFilter===2 ? true : false}                       
            />
        </div>
    );
}
// default props value
ToDoFilter.defaultProps = {
    handleOnClick: () => {},
    activeFilter: 0
  };
ToDoFilter.propstypes = {
    handleOnClick: PropTypes.func.isRequired,
    activeFilter: PropTypes.number.isRequired
}
export default ToDoFilter;