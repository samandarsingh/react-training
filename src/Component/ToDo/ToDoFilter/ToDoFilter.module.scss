.todoFilterContainer {
    margin-top: 50px;
}