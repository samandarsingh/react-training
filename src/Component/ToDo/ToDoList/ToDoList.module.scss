.todoList {       
    width: 50%;
    table {
        margin-top: 10px;
        width: 100%;
        font-family: "Trebuchet MS", Arial, Helvetica, sans-serif;
        border-collapse: collapse;
    }    
    th, td {
        border: 1px solid #ddd;
        text-align: left;
        padding: 8px;
    }
    th {
        padding-top: 12px;
        padding-bottom: 12px;
        text-align: left;
        background-color: #4c8aaf;
        color: white;
    }      
}