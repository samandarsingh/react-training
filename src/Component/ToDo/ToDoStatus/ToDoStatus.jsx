import React from "react";
import PropTypes from "prop-types";
import Button from "../../Shared/Button";
/**
 * this component renders status button
 * @param {bool} toDoStatus 
 * @param {func} handleOnClick\
 * @param {number} index
 * @returns Element
 */
const ToDoStatus = props => {
    const {
        toDoStatus,
        handleOnClick,
        index
    } = props;
    return (
        <Button 
            buttonType="button"
            buttonName={toDoStatus ? "Active":"Decline"}
            buttonClass={toDoStatus ? "statusButton":"declineButton"}
            handleOnClick={handleOnClick}
            index={index}
            status={toDoStatus ? true:false}            
        />
    );
}
ToDoStatus.defaultProps = {
    toDoStatus: true,
    handleOnClick: () => {},
    index: 0
}
ToDoStatus.propTypes = {
    toDoStatus: PropTypes.bool.isRequired,
    handleOnClick: PropTypes.func.isRequired,
    index: PropTypes.number.isRequired
}
export default ToDoStatus;