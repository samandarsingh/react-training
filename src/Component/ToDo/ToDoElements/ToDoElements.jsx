import React from "react";
import PropTypes from "prop-types";
import Text from "../../Shared/Text";
import ToDoStatus from "../ToDoStatus";
import CompleteToDoStatus from "../CompleteToDoStatus";
import ToDoAction from "../ToDoAction";
/**
 * 
 * This function for ToDo Elements row in table
 * @param {number} index
 * @param {string} toDoTitle
 * @param {bool} toDoStatus
 * @param {bool} completeToDoStatus
 * @param {func} updateToDoStatus
 * @param {func} updateToDoCompleteStatus
 * @param {func} updateToDo
 * @param {func} removeToDo  
 */
const ToDoElements = props => {
    const {
        index,
        toDoTitle,
        toDoStatus,
        completeToDoStatus,
        updateToDoStatus,
        updateToDoCompleteStatus,
        updateToDo,
        removeToDo
    } = props;
    return (
        <tr>                                
            <td><Text title={toDoTitle} /></td>
            <td>
                <ToDoStatus 
                  toDoStatus={toDoStatus}  
                  handleOnClick={updateToDoStatus}
                  index={index}
                />
            </td>
            <td>
                <CompleteToDoStatus
                    completeToDoStatus={completeToDoStatus}
                    handleOnClick={updateToDoCompleteStatus}
                    index={index}
                />                
            </td>
            <td>
                <ToDoAction
                    handleOnClickUpdateToDoStatus={updateToDo}
                    handleOnClickRemoveToDoStatus={removeToDo}
                    index={index}
                /> 
            </td>
        </tr>
    );
}
ToDoElements.defaultProps = {
    index: 0,
    toDoTitle: "",
    toDoStatus: true,
    CompleteToDoStatus: false,
    updateToDoStatus: () => {},
    updateToDoCompleteStatus: () => {},
    updateToDo: () => {},
    removeToDo: () => {}
}
ToDoElements.propstypes = {
    index: PropTypes.number.isRequired,
    toDoTitle: PropTypes.string.isRequired,
    toDoStatus: PropTypes.bool.isRequired,
    completeToDoStatus: PropTypes.bool.isRequired,
    updateToDoStatus: PropTypes.func.isRequired,
    updateToDoCompleteStatus: PropTypes.func.isRequired,
    updateToDo: PropTypes.func.isRequired,
    removeToDo: PropTypes.func.isRequired
}
export default ToDoElements;