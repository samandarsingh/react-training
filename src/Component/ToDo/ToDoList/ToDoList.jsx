import React from "react";
import PropTypes from "prop-types";
import styles from "./ToDoList.module.scss";
import ToDoElements from "../ToDoElements";
import ToDoFilter from "../ToDoFilter";
/**
 * This component for render todo list
 * @param {array} toDos // array of objects
 */
class ToDoList extends React.Component {
    constructor(props){
        super(props);
        this.state = {
            toDos: this.props.toDos,
            activeFilter: 0,
            todosArr :[]
        }        
    }
    componentDidMount(){        
        this.TodoFilter("",0);
    }
    //receive props
    componentDidUpdate(prevProps, prevState) {
        if (this.props.toDos.length !== prevState.toDos.length) { 
            this.setState({
                toDos: prevProps.toDos
            },()=>{
                this.TodoFilter("",0);
            });                                  
        }
    }
    // update todo
    updateTodo = (e, index) => { 
        this.props.handleOnClick(e,index);        
    };         
    // remove todo
    removeTodo = (todo) =>{
        const {toDos,updateToDos} = this.props;
        const affterDelete = toDos.filter(item => {
            return item.todo !== todo;
        });
        updateToDos(affterDelete);
    }  
    // update todo status
    updateTodoStatus = (e, index, status) => {        
        let newTodoDataArray = [...this.state.toDos]; 
        for(let i = 0; i < newTodoDataArray.length;i++){
            if(i===index){
                newTodoDataArray[i].status = !status;
            }
        }
        this.setState({
            toDos: newTodoDataArray
        },() => {
            this.TodoFilter("",0);
        });
    }
    // update todo complete status
    updateTodoCompleteStatus = (e, index, status) => {       
        let newComTodoDataArray = [...this.state.toDos]; 
        for(let i = 0; i < this.state.toDos.length;i++){
            if(i===index){
                newComTodoDataArray[i].isCompleted = !status;
            }
        }
        this.setState({
            toDos: newComTodoDataArray
        });
    }
    //todo filter
    TodoFilter = (e, filter) => {        
        let newFilterArr = [...this.state.toDos];
        let filterData = [];
        if (filter === 0){
            // filter get record all
            filterData = newFilterArr;
        } else if (filter === 1){
            // filter get active record
            filterData=this.props.toDos.filter(function(todo){
                return todo.status !== false;                
            });
        } else if (filter === 2){
            // filter get is completed record
            filterData = this.props.toDos.filter(function(todo){
                return todo.isCompleted !== false;
            });
        }        
        this.setState({
            todosArr: filterData,
            activeFilter: filter
        });
    }
    render () {
        const {activeFilter,todosArr} = this.state;        
        return (            
            <div className={styles.todoList}>
                <ToDoFilter 
                handleOnClick={this.TodoFilter}
                activeFilter={activeFilter}/>
                <table>
                    <thead>
                        <tr>
                            <th>Todo</th>
                            <th>Status</th>
                            <th>Is Completed</th>
                            <th>Action</th>
                        </tr>                        
                    </thead>
                    <tbody>                         
                        {todosArr.map((item,itemKey) => {
                            return (
                                <ToDoElements 
                                index={itemKey}
                                toDoTitle={item.todo}
                                toDoStatus={item.status}
                                completeToDoStatus={item.isCompleted}
                                updateToDoStatus={this.updateTodoStatus}
                                updateToDoCompleteStatus={this.updateTodoCompleteStatus}
                                updateToDo={this.updateTodo}
                                removeToDo={()=>this.removeTodo(item.todo)}
                                key={itemKey}
                                />                            
                            );
                        })
                        } 
                    </tbody>
                </table>
            </div>            
        ); 
    };
}
ToDoList.defaultProps = {
    updateToDos: () => {},
    toDos: [],
    handleOnClick: () => {}
}
ToDoList.propTypes = {
    toDos: PropTypes.array,
    updateToDos: PropTypes.func.isRequired,
    handleOnClick: PropTypes.func.isRequired
}
export default ToDoList;