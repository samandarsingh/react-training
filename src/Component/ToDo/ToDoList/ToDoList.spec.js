import React from "react";
import { mount } from "enzyme";
import ToDoList from "./ToDoList";
import ToDoFilter from "../ToDoFilter";
import Button from "../../Shared/Button";
import ToDoStatus from "../ToDoStatus";
import CompleteToDoStatus from "../CompleteToDoStatus";
import ToDoElements from "../ToDoElements";
import ToDoAction from "../ToDoAction";
describe(`ToDoList component testing`, () => {
    let wrapper = "";
    beforeAll(() => {
        const props  = {
            toDos: [{todo:"test",status:true,isCompleted:false}]
        }
        wrapper = mount(<ToDoList {...props} />);
    });
    it(`ToDoList component should be render without error`, () => {
        expect(wrapper).toBeTruthy();
    });
    it(`Testing ToDoFilter button class click on that button class should be there`, () => {
        //console.log(wrapper.debug());
        wrapper
        .find(ToDoFilter)
        .find(`button`)
        .at(1)
        .simulate(`click`);
        expect(
            wrapper
            .find(ToDoFilter)
            .find(`button`)
            .at(0)
            .hasClass(`statusButton`)            
        ).toEqual(true);
        expect(            
            wrapper
            .find(ToDoFilter)
            .find(`button`)
            .at(1)
            .hasClass(`isCompletedButton`)            
        ).toEqual(true);
        expect(
            wrapper
            .find(ToDoFilter)
            .find(`button`)
            .at(2)
            .hasClass(`statusButton`)
        ).toEqual(true);        
    });
    it(`Testing table th mock data`, () => {
        expect(
            wrapper
            .find(`table`)
            .find(`thead`)
            .find(`tr`)
            .find(`th`)
            .at(0)
            .text()
        ).toEqual(`Todo`);
        expect(
            wrapper
            .find(`table`)
            .find(`thead`)
            .find(`tr`)
            .find(`th`)
            .at(1)
            .text()
        ).toEqual(`Status`);
        expect(
            wrapper
            .find(`table`)
            .find(`thead`)
            .find(`tr`)
            .find(`th`)
            .at(2)
            .text()
        ).toEqual(`Is Completed`);
        expect(
            wrapper
            .find(`table`)
            .find(`thead`)
            .find(`tr`)
            .find(`th`)
            .at(3)
            .text()
        ).toEqual(`Action`);
    });
    it(`testing on click to do status button class should be change`, () => {        
        wrapper
        .find(ToDoStatus)
        .find(`button`)
        .simulate(`click`);
        expect(
            wrapper
            .find(ToDoStatus)
            .find(`button`)
            .hasClass(`declineButton`)
        ).toEqual(true);
        wrapper
        .setState({
            toDos: [{todo:"test",status:true,isCompleted:false}]
        });
        expect(
            wrapper
            .find(ToDoElements)            
        ).toHaveLength(1);
    });
    it(`testing on click to do complete status button class should be change`, () => {         
        wrapper
        .find(ToDoFilter)
        .find(`button`)
        .at(2)
        .simulate(`click`);
        expect(
            wrapper
            .find(ToDoFilter)
            .find(`button`)
            .at(2)
            .hasClass(`isCompletedButton`)
        ).toEqual(true);        
    });
    it(`testing all on click change button class should`, () => {
        wrapper
        .find(ToDoFilter)
        .find(`button`)
        .at(0)
        .simulate(`click`)
        expect(
            wrapper
            .find(ToDoFilter)
            .find(`button`)
            .at(0)
            .props()  
            .disabled          
        ).toEqual(true);
        expect(
            wrapper
            .find(ToDoFilter)
            .find(`button`)
            .at(0)
            .hasClass(`isCompletedButton`)
        ).toEqual(true);
    });
    it(`testing Active on click change button class should`, () => {
        wrapper
        .find(ToDoFilter)
        .find(`button`)
        .at(1)
        .simulate(`click`)
        expect(
            wrapper
            .find(ToDoFilter)
            .find(`button`)
            .at(1)
            .props()  
            .disabled          
        ).toEqual(true);
        expect(
            wrapper
            .find(ToDoFilter)
            .find(`button`)
            .at(1)
            .hasClass(`isCompletedButton`)
        ).toEqual(true);
    });
    it(`testing is completed on click change button class should`, () => {
        wrapper
        .find(ToDoFilter)
        .find(`button`)
        .at(2)
        .simulate(`click`)
        expect(
            wrapper
            .find(ToDoFilter)
            .find(`button`)
            .at(2)
            .props()  
            .disabled          
        ).toEqual(true);
        expect(
            wrapper
            .find(ToDoFilter)
            .find(`button`)
            .at(2)
            .hasClass(`isCompletedButton`)
        ).toEqual(true);
    });
    it(`SnapShot Testing`, () => {
        expect(wrapper).toMatchSnapshot();
    });
});