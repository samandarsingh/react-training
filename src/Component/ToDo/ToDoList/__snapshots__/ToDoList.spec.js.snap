// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDoList component testing SnapShot Testing 1`] = `ReactWrapper {}`;
